File: frontend/kube-guard-frontend/src/api/ops.ts
// src/api/ops.ts
import { http } from '@/api/http'

export interface HealEvent {
  id: number
  ts: number
  namespace: string
  pod: string
  reason: string
  action: string
  result: string
}

export interface OpsAction {
  id: number
  ts: number
  action: string
  target: string
  params: string
  dry_run: number
  result: string
  detail: string
}

export interface HealStatusResp {
  running: boolean
  interval_sec: number
  last_run_ts: number | null
  last_summary: any | null
  last_error: string | null
  enabled: boolean
  execute: boolean
  max_per_cycle: number
  cooldown_sec: number
  allow_ns: string
  deny_ns: string
  only_reasons: string
}

export function runHealOnce(params?: { namespace?: string }) {
  return http.post('/api/ops/heal/run', null, { params })
}

export function fetchHealEvents(params?: { limit?: number; offset?: number }) {
  return http.get<{ items: HealEvent[] }>('/api/ops/heal/events', { params })
}

export function deleteHealEvent(id: number) {
  return http.delete<{ ok: boolean }>('/api/ops/heal/events/' + id)
}

export function fetchOpsActions(params?: { limit?: number; offset?: number }) {
  return http.get<{ items: OpsAction[] }>('/api/ops/actions', { params })
}

export function deleteOpsAction(id: number) {
  return http.delete<{ ok: boolean }>('/api/ops/actions/' + id)
}

export function fetchHealStatus() {
  return http.get<HealStatusResp>('/api/ops/heal/status')
}

// ===================== ✅ 阶段 3.3：Deployment 视图 =====================

export type HealDeploymentStatus = 'normal' | 'pending' | 'circuit_open'

export interface HealDeploymentItem {
  namespace: string
  deployment_uid: string
  deployment_name: string
  status: HealDeploymentStatus
  fail_count: number
  is_failing: number
  pending_until_ts: number
  last_action: string
  last_reason: string
  last_pod: string
  last_ts: number
  last_result: string
}

export function fetchHealDeployments(params?: { limit?: number; offset?: number; namespace?: string }) {
  return http.get<{ items: HealDeploymentItem[] }>('/api/ops/heal/deployments', { params })
}

export function fetchHealDeploymentDetail(namespace: string, deployment_uid: string) {
  return http.get<{ ok: boolean; item?: HealDeploymentItem; reason?: string }>(
    `/api/ops/heal/deployments/${namespace}/${deployment_uid}`
  )
}

// ===================== ✅ 复位（reset） =====================

export interface HealResetReq {
  namespace?: string | null
  deployment_uid?: string | null
  deployment_name?: string | null
  restore_replicas?: number | null
}

export interface HealResetResp {
  ok: boolean
  reason?: string | null
}

export function healReset(payload: HealResetReq) {
  return http.post<HealResetResp>('/api/ops/heal/reset', payload)
}

// ===================== ✅ Decay 配置（按钮用） =====================

export interface HealDecayCfgResp {
  enabled: boolean
  step: number
}

export function fetchHealDecayConfig() {
  return http.get<HealDecayCfgResp>('/api/ops/heal/decay')
}

export function setHealDecayConfig(payload: HealDecayCfgResp) {
  return http.put<HealDecayCfgResp>('/api/ops/heal/decay', payload)
}

File: backend/kube-devops-backend/fastApiProject/routers/ops.py
# routers/ops.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from services.ops.heal_reset import reset_heal_state
from services.ops.schemas import HealResetReq, HealResetResp
from services.ops.audit import list_events, delete_event_by_id, list_actions, delete_action_by_id
from services.ops.healer import run_heal_scan_once
from services.ops.scheduler import get_status
from services.ops.heal_view import list_heal_deployments, get_heal_deployment_detail
from services.alert_client import list_alerts
from services.ops.runtime_config import get_heal_decay_config, set_heal_decay_config

router = APIRouter(prefix="/api/ops", tags=["Ops"])


# -------------------- heal events --------------------
@router.get("/heal/events")
def heal_events(limit: int = 50, offset: int = 0):
    return {"items": list_events(limit=limit, offset=offset)}


@router.delete("/heal/events/{event_id}")
def heal_event_delete(event_id: int):
    try:
        ok = delete_event_by_id(event_id)
        if not ok:
            raise HTTPException(status_code=404, detail="event not found")
        return {"ok": True}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"delete event failed: {e}")


# -------------------- heal run/reset/status --------------------
@router.post("/heal/run")
def heal_run_once(namespace: Optional[str] = None):
    try:
        result = run_heal_scan_once(namespace=namespace)
        return {"ok": True, "result": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"heal run failed: {e}")


@router.get("/heal/status")
def heal_status():
    # 你原来就是直接 return get_status()（:contentReference[oaicite:6]{index=6}）
    return get_status()

@router.post("/heal/reset", response_model=HealResetResp)
def heal_reset(req: HealResetReq):
    out = reset_heal_state(
        namespace=req.namespace,
        deployment_uid=req.deployment_uid,
        deployment_name=req.deployment_name,
        restore_replicas=req.restore_replicas,
    )
    if not out.get("ok"):
        raise HTTPException(status_code=400, detail=out.get("reason") or "reset failed")
    return HealResetResp(**out)


# -------------------- ✅ heal decay config (for UI buttons) --------------------
class HealDecayCfgResp(BaseModel):
    enabled: bool
    step: int


class HealDecayCfgReq(BaseModel):
    enabled: bool = Field(..., description="enable decay on recovered")
    step: int = Field(1, ge=1, le=10, description="decay step per recovered")


@router.get("/heal/decay", response_model=HealDecayCfgResp)
def heal_decay_get():
    enabled, step = get_heal_decay_config()
    return HealDecayCfgResp(enabled=enabled, step=step)


@router.put("/heal/decay", response_model=HealDecayCfgResp)
def heal_decay_set(req: HealDecayCfgReq):
    enabled, step = set_heal_decay_config(enabled=req.enabled, step=req.step)
    return HealDecayCfgResp(enabled=enabled, step=step)


# -------------------- actions audit --------------------
@router.get("/actions")
def actions(limit: int = 50, offset: int = 0):
    return {"items": list_actions(limit=limit, offset=offset)}


@router.delete("/actions/{action_id}")
def action_delete(action_id: int):
    try:
        ok = delete_action_by_id(action_id)
        if not ok:
            raise HTTPException(status_code=404, detail="action not found")
        return {"ok": True}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"delete action failed: {e}")


# -------------------- ✅ 阶段3.3：Deployment 维度聚合视图 --------------------
@router.get("/heal/deployments")
def heal_deployments(limit: int = 200, offset: int = 0, namespace: Optional[str] = None):
    """
    给前端 Tab 展示用：正常 / pending / 熔断中 + 最近一次自愈/失败次数
    """
    return {"items": list_heal_deployments(limit=limit, offset=offset, namespace=namespace)}


@router.get("/heal/deployments/{namespace}/{deployment_uid}")
def heal_deployment_detail(namespace: str, deployment_uid: str):
    d = get_heal_deployment_detail(namespace=namespace, deployment_uid=deployment_uid)
    if not d:
        raise HTTPException(status_code=404, detail="deployment not found")
    return d


# -------------------- ✅ 告警（Alertmanager 当前 alerts） --------------------
@router.get("/alerts")
def alerts():
    return {"items": list_alerts()}